service/resource/resourcegroup/resource.go

```diff
@@ -0,0 +1,289 @@
+package resourcegroup
+
+import (
+	"time"
+
+	azureresource "github.com/Azure/azure-sdk-for-go/arm/resources/resources"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/giantswarm/azuretpr"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"github.com/giantswarm/operatorkit/framework"
+
+	"github.com/giantswarm/azure-operator/client"
+	"github.com/giantswarm/azure-operator/service/key"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "resourcegroup"
+
+	clusterIDTag  = "ClusterID"
+	customerIDTag = "CustomerID"
+	deleteTimeout = 5 * time.Minute
+	managedBy     = "azure-operator"
+)
+
+type Config struct {
+	// Dependencies.
+
+	AzureConfig *client.AzureConfig
+	Logger      micrologger.Logger
+}
+
+// DefaultConfig provides a default configuration to create a new resource by
+// best effort.
+func DefaultConfig() Config {
+	return Config{
+		// Dependencies.
+		AzureConfig: nil,
+		Logger:      nil,
+	}
+}
+
+type Resource struct {
+	// Dependencies.
+
+	azureConfig *client.AzureConfig
+	logger      micrologger.Logger
+}
+
+// New creates a new configured resource group resource.
+func New(config Config) (*Resource, error) {
+	// Dependencies.
+	if config.AzureConfig == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.AzureConfig must not be empty.")
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.Logger must not be empty.")
+	}
+
+	newService := &Resource{
+		azureConfig: config.AzureConfig,
+		logger: config.Logger.With(
+			"resource", Name,
+		),
+	}
+
+	return newService, nil
+}
+
+// GetCurrentState gets the resource group for this cluster from the Azure API.
+func (r *Resource) GetCurrentState(obj interface{}) (interface{}, error) {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var group *Group
+	{
+		groupsClient, err := r.getGroupsClient()
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		resourceGroup, err := groupsClient.Get(key.ClusterID(customObject))
+		if err != nil {
+			if client.ResponseWasNotFound(resourceGroup.Response) {
+				// Fall through.
+				return &Group{}, nil
+			}
+
+			return nil, microerror.Mask(err)
+		}
+		group = &Group{
+			Name:     *resourceGroup.Name,
+			Location: *resourceGroup.Location,
+			Tags:     to.StringMap(*resourceGroup.Tags),
+		}
+	}
+
+	return group, nil
+}
+
+// GetDesiredState returns the desired resource group for this cluster.
+func (r *Resource) GetDesiredState(obj interface{}) (interface{}, error) {
+	customObject, err := toCustomObject(obj)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	tags := map[string]string{
+		clusterIDTag:  key.ClusterID(customObject),
+		customerIDTag: key.ClusterCustomer(customObject),
+	}
+	resourceGroup := &Group{
+		Name:     key.ClusterID(customObject),
+		Location: key.Location(customObject),
+		Tags:     tags,
+	}
```

Comment:
Why we pass group as a pointer if we still check if `Name` is empty?

Reply:
We return nil on error or an empty group if there is no resource group. Using the pointer means we can differentiate between these.

Reply:
We should detect if we have error by checking if err is nil. I don't see problem with returning zero value when error is not nil.

Reply:
Cool, I'll get rid of the pointer.